### DISCUSSION TITLE
Not understanding propagation

### DISCUSSION THREAD

**Discussion by thelitbulb** [markqvist/Reticulum#139] [Sideband]
I finally got a Nomadnet node running on my VPS and conversations between Sideband and the server work fine.  I'm a little confused as to how propagation works.  In Sideband whenever I click the top right button it switches from propagation mode to direct mode?  I guess?  When in propagation mode (icon has an up arrow) my outgoing messages (from sideband to server) have the "On Propagation Net" state which never clears.  The server receives the messages but for some reason the ack isn't making it back to Sideband.

I'm also a little confused by the "LXMF Propagation Node" setting under Preferences.  I put the ID of the propagation node on my server but what does this feature actually do?

Thanks in advance and keep up the good work this project is world changing!

**Comment by markqvist** [comment]
The documentation for this is a bit scattered and hard to find right now, so I can definitely understand your confusion. Doing the actual LXMF documentation is on my long list of things (but pretty high up, at least). In the meantime, the best explanation comes from the **Guide** section in the `nomadnet` program. While the below talks about the `nomadnet` program, the same is true for Sideband:

> By default, Nomad Network will attempt to deliver messages to a peer directly. This happens by first establishing an encrypted link directly to the peer, and then delivering the message over it.
> 
> If the desired peer is not available because it has disconnected from the network, this method will obviously fail. In this case, Nomad Network will attempt to deliver the message to a node, which will store and forward it over the network, for later retrieval by the destination peer. The message is encrypted before being transmitted to the network, and is only readable by the intended recipient.
> 
> For propagated delivery to work, one or more nodes must be available on the network. If one or more trusted nodes are available, Nomad Network will automatically select the most suitable node to send the message via, but you can also manually specify what node to use.
> 
> Nodes also form a distributed message store for offline users, and allows messages to be exchanged between peers that are not online at the same time.
> 
> By default, Nomad Network will check in with propagation nodes, and download any available messages periodically. You can change this by editing the configuration.
> 
> If no propagation nodes exist on a network, all peers will still be able to communicate directly peer-to-peer, but both endpoints of a conversation will need to be available at the same time to converse.
> 
> When nodes exist on the network, messages will be held and syncronised between nodes for deferred delivery if the destination peer is unavailable. Nodes will automatically discover and peer with each other, and handle syncronisation of message stores between them.

I just copy and pasted the parts I thought would make sense, so feel free to ask for clarification if you want to know more!

**Comment by SebastianObi** [comment]
Thank you for the explanation,

So far it is also logical and understandable. It also works as described. Only one thing is still unclear to me:

 `When nodes exist on the network, messages will be held and syncronised between nodes for deferred delivery if the destination peer is unavailable. Nodes will automatically discover and peer with each other, and handle syncronisation of message stores between them.`

So when and how are the messages synchronized between the propagation nodes? I have tried to test this (2 nodes) but no messages can be retrieved from the 2nd node.

Maybe you can give me a short answer (it doesn't have to be a detailed explanation). Only what is necessary for this to work and how I can test it.

Thanks

BTW: Is it possible to define the display_name of the propagation node? I have not found anything in the NomadNet configuration. That would be cool if that would work too. At the moment the annoucne stream shows "Propagation Node" for every node.

**Reply by markqvist** [reply]
> So when and how are the messages synchronized between the propagation nodes? I have tried to test this (2 nodes) but no messages can be retrieved from the 2nd node.

At the moment, all that is needed for a Propagation Node to peer with another one is:

- To have auto-peering enabled on the Propagation Node
- To hear an announce from another Propagation Node, that is not further away than the `peering_maxdepth` property.
- When the above happens, the node is automatically peered immediately.
- When a node has added a new peer, it will start syncing messages to that peer that arrives *after the first peering time*. Messages that existed in the message store *before* the peering happened will not be synced.
- Propagation Nodes will not get to know about messages that existed on the network *before they became active*. This is by design.

> Is it possible to define the display_name of the propagation node? I have not found anything in the NomadNet configuration.

Currently no, and I don't *think* it will be. But let's see a little further down the line, for now it is just as is, though. The core of the LXMF Propagation protocol should stay as lightweight as entirely possible, so adding a lot of data to the announces is not too attractive, I think.

But what you can do is just retrieve the name of another associated service, as Nomad Network does. It simply uses the nomadnet node name associated with the same Reticulum Identity. Just reconstruct the destination hash for how the named destination, and query the announce data for that, then you can associate that with the name in your app.

**Reply by SebastianObi** [reply]
Ok, when I have time I will try this out.
As you said it makes no sense to change the protocol and transport the display_name. I was just confused because in NomadNet the name was displayed and in Sideband it was not. Reconstructing the name in the app is a good solution.

**Comment by thelitbulb** [comment]
I have since rebuilt the server and I've learned a lot. I switched to Nomadnet as a client. The usability of Sideband is... rough.

As far as the main question I had with Sideband, it had to do with propagation acknowledgements in the app. It's probably just a bug in the app and not an actual issue with Reticulum. It's no longer an issue since I'm now using only Nomadnet. : )

I ran into some other issues during my adventures maybe you guys have some insight.

The node (vps) always says "Currently, no nodes are known" in the "Saved Nodes" window. I have share_instance enabled and when I announce from the node it never appears in the announce stream. Do nodes ignore their own announce messages? Is it a problem that my node doesn't recognize itself as a node? I can see my index.mu from itself (using browse) and also from the clients.

I noticed that when I change the name in my client or on the node and then enter a conversation that I get a triangle exclamation glyph next to the name. Even when the ID is trusted. If I change the name back to the old name in the conversations pane then it goes back to all green. My take? I think after you have an initial announce Nomadnet doesn't like it when you change names... potentially to prevent spoofing? It's slightly inconvenient not being able to change your name at all though.

On the node in the log it always says "Could not autoselect a propagation node! LXMF propagation will not be available until a trusted node announces on the network." just before "Starting user interface...". Is this related to 1? Or have I misconfigured something?

EDIT: Is a propagation node absolutely necessary for store and forward? Say for example I have one node and two clients with direct peer connections to that node. Client A sends a message to client B while it is offline. Would the message be eventually received by client B? Or would a propagation node need to be configured first?

EDIT: Is it needed or useful for me to offer my node for public use? If so, how would I go about advertising that to others?

Thanks again you guys! Doing gods work : )

EDIT:  Just realized I posted this in the wrong place.  Fixed.

**Comment by markqvist** [comment]
Sorry for being a little slow in getting back to you here!

> I announce from the node it never appears in the announce stream. Do nodes ignore their own announce messages? 

Yes, the node will ignore it's own announces in the announce stream. You will only see announces from other nodes and persons, along with standalone LXMF propagation nodes (systems that just run a message propagation node through the `lxmd` program, but not a full Nomad Network node).

> I noticed that when I change the name in my client or on the node and then enter a conversation that I get a triangle exclamation glyph next to the name. Even when the ID is trusted. If I change the name back to the old name in the conversations pane then it goes back to all green.

Thanks. Yeah, there is currently something a little bit too aggressive with this logic. It should of course be possible to change the name, that is definitely the point ;) It was made like that to help people spot if someone is trying to act as a contact you already have saved in your list, and *another* address announces the same name, but it is obviously a bit over the top now. I should have it improved in the next release.

> Is a propagation node absolutely necessary for store and forward? Say for example I have one node and two clients with direct peer connections to that node. Client A sends a message to client B while it is offline. Would the message be eventually received by client B? Or would a propagation node need to be configured first?

Yes, there must be at least one propagation node somewhere on your network for store-and-forward functionality to work. A propagation node *is* basically just that, as per default, the design philosophy of Reticulum and related programs is not do run or do anything the user itself did not ask for. So it has to be enabled willfully by someone that actually wants to run it.

If you don't want to run a full `nomadnet` node, you can just use the much lighter `lxmd` to run a propagation-only node. Run `lxmd --help` and `lxmd --exampleconfig` to learn how to set it up.

> Thanks again you guys! Doing gods work : )

Thank you. I am glad you are finding it useful.

**Reply by markqvist** [reply]
Also, let me know if I missed any of your questions, or if you need more info on anything!

**Reply by markqvist** [reply]
Also, also: It really helps reading the entire **Guide** section inside the `nomadnet` program itself, if you didn't already. It contains all of the background info needed to understand how the system works.
